bson/bsoncodec/map_codec.go

```diff
@@ -22,6 +24,7 @@ var defaultMapCodec = NewMapCodec()
 type MapCodec struct {
 	DecodeZerosMap   bool
 	EncodeNilAsEmpty bool
+	MgoKeyHandling   bool
```

Comment:
I'd prefer to contain any named references to `mgo` to the `mgocompat` package and have these options be generically named, but set appropriately to create the `mgocompat.Registry` instead. Can we rename this to something like `UseStringerForKeys`?

Reply:
Call it `EncodeKeysWithStringer` to maych the `EncodeNilAsEmpty`

I'm still wondering why we can't take a `func(reflect.Value) (string, error)` here...  `KeyEncoder` and then have a `KeyDecoder` as well.

Reply:
This keeps parity with our other interfaces (Marshaler/ValueMarshaler/Proxy/etc). I also prefer to avoid exposing the `reflect` package if we can. I know our overall codec system is built on it, but if we can get away with a simpler function, that's my preference. Also, if we changed this to explicitly take a function, this behavior would be opt-in rather than opt-out and using `Stringer` would be the default, which we've agreed is broken.

There is one remaining discussion: do we want to take `encoding.TextMarshaler` from the standard library's `encoding` package or make our own interface/function? Do we expect a user to want different behavior when using stdlib json vs our BSON library?

Reply:
fair enough.